ui/app/src/components/Navigation/FolderBrowserTreeView.tsx

```diff
@@ -0,0 +1,281 @@
+  createStyles({
+    wrapper: {
+      minHeight: '60vh',
+      display: 'flex',
+      flexDirection: 'column'
+    },
+    loadingState: {
+      flexGrow: 1,
+      flexDirection: 'unset'
+    },
+    loading: {
+      display: 'flex',
+      alignItems: 'center',
+      padding: '2px',
+      marginLeft: '5px',
+      '& p': {
+        marginRight: '10px'
+      }
+    },
+    treeItemRoot: {
+      '&:focus > $treeItemContent, &$treeItemSelected > $treeItemContent': {
+        color: `${palette.blue.main}`
+      }
+    },
+    treeItemContent: {},
+    treeItemSelected: {},
+    treeItemLabel: {
+      width: 'auto',
+      padding: '0 4px',
+      borderRadius: '5px'
+    }
+  })
+);
+
+export function legacyItemsToTreeNodes(items: LegacyItem[]) {
+  return items
+    .filter((item) => item.contentType === 'folder')
+    .map((item) => ({
+      id: item.path,
+      name: item.name,
+      children: [
+        {
+          id: 'loading'
+        }
+      ]
+    }));
+}
+
+interface FolderBrowserTreeViewProps {
+  rootPath: string;
+  currentPath: string;
+  expanded: string[];
+  selected: string;
+  invalidPath: boolean;
+  resource: Resource<TreeNode>;
+  classes?: Partial<Record<'treeViewRoot', string>>;
+  onNodeSelected(event: React.ChangeEvent<{}>, nodeId: string): void;
+  onNodeToggle(event: React.ChangeEvent<{}>, nodeIds: string[]): void;
+  onPathChanged(path: string): void;
+}
+
+export default function FolderBrowserTreeView(props: FolderBrowserTreeViewProps) {
+  const classes = useStyles({});
+  const {
+    rootPath,
+    currentPath,
+    expanded,
+    selected,
+    onNodeToggle,
+    onNodeSelected,
+    resource,
+    onPathChanged,
+    invalidPath
+  } = props;
+
+  const treeNodes = resource.read();
+
+  return (
+    <section className={classes.wrapper}>
+      <PathSelected
+        rootPath={rootPath}
+        currentPath={currentPath.replace(rootPath, '')}
+        onPathChanged={onPathChanged}
+        invalidPath={invalidPath}
+      />
+      {
+        treeNodes ? (
+          <TreeView
+            classes={{ root: props.classes?.treeViewRoot }}
+            defaultCollapseIcon={<ExpandMoreIcon />}
+            defaultExpandIcon={<ChevronRightIcon />}
+            expanded={expanded}
+            selected={selected}
+            onNodeToggle={onNodeToggle}
+            onNodeSelect={onNodeSelected}
+          >
+            <RenderTreeNode node={treeNodes} />
+          </TreeView>
+        ) : (
+          <LoadingState classes={{ root: classes.loadingState }} />
+        )
+      }
+    </section>
+  );
+}
+
+interface RenderTreeNode {
+  node: TreeNode;
+}
+
+function RenderTreeNode({ node }: RenderTreeNode) {
+  const classes = useStyles({});
+  return node.id === 'loading' ? (
+    <div className={classes.loading}>
+      <Typography>
+        <FormattedMessage id="folderBrowserTreeView.loading" defaultMessage={'Loading...'} />
+      </Typography>
+      <CircularProgress size={16} />
```

Comment:
Looks like this is still inverted. The spinner should be to the left of the message.

Use \<FormattedMessage id=words.loading/>&hellip;